### ISSUE TITLE
Iterable list is complained to be an object

### ISSUE THREAD

**Issue by Haiz14** [microsoft/pyright#7275]
**Describe the bug**
On iterating over a list of values in a dict, Pyright complains
` │     "object" is not iterable   "__iter__" method not defined pyright (reportGeneralTypeIssues) [16, 20]
`

**Code or Screenshots**
[pyright-play.net](https://pyright-play.net/?code=GYJw9gtgBALgngBwJYDsDmUkQWEMoAySAzjADRQAqiApgCYAiSAxjAFBvMA2AhscVACyAVy4wkCLjQBKPAO4NIPVAAVwdYawY8YPJqwAU1BPX0wAlAC42UW1ADEhEvjDAoIeVDo6eNuwFUAZUsnUgBtOhYYAF0oAB8oADkwFBo-WwBhAEEQonDI1liE5NT0qABxACFc5wiooqSUtLsoBgBRGvz6%2BMbStjoaNwQ4ECQ0AAsYAH0AI2E0A29dEJExCSlZBSVVdU0YbV0zcygAWgA%2BXpprFuBcKABrGjgKADceLmEaTBQvHwA6JAwGgQYgGKxlFpINxvD5fEiXELMFLiFCfKAQuyOYajCb4NBIF40AQ0EDgEBQFKYIEeFEYMCE8kAAxhn0ZGNsjgARGAZgArGisTmYAQoMD4QEkngzKRQAAFss5UymEpASqFEBoMHGYDoUFF%2BAGwFQ9Cg2LGkygBhANBweHKNFSHi4xhoAEl%2BJ9iMcwgBGADMFAATAAGaIcFo3O4w75QFlXdkRhCjFAwAww8xsIA)

```python
from typing import List, TypedDict

class MultipleRawDomainProductDataDict(TypedDict):
    # List of raw data
    US: List[dict] | None
    CA: List[dict] | None
    GB: List[dict] | None
    DE: List[dict] | None

def pyright_bug(data: MultipleRawDomainProductDataDict) -> None:
    for key, value in data.items():
        if value is None: continue 
        # pyright gives error on iterating over `value`
        # "object" is not iterable   "__iter__" method not defined pyright (reportGeneralTypeIssues) [13, 20]

        for val in value:
            print(val)

```
![Screenshot_2024-02-16-19-44-18-28_84d3000e3f4017145260f7618db1d683](https://github.com/microsoft/pyright/assets/68465021/4c1ff176-1f99-43a7-9443-d73f5414a902)

**VS Code extension or command-line**
Pyright lsp (neovim)  
```bash
❯ pyright --version
pyright 1.1.350

```

**Comment by erictraut** [comment]
Pyright's behavior here is correct. A `TypedDict` is a structural type, not a nominal class. It describes any `dict` or `Mapping` that has the specified keys with the specified types, but it doesn't preclude there from being additional keys present. That means the type checker must assume there will be additional keys, and their types cannot be known, other than saying they are type `object`.

There is a draft [PEP 728](https://peps.python.org/pep-0728/) currently under review and discussion that would allow a `TypedDict` to be declared as "closed" such that it cannot have any additional items.

**Comment by Haiz14** [comment]
Alternative if anyone see's this
```py
MultipleRaw___ = dict[Literal["US", ___], list[dict] | None]
```
